src/main/host/descriptor/pipe.rs

```diff
@@ -301,26 +310,43 @@ impl PipeFile {
         self.state
     }
 
-    fn filter_state(&self, mut state: FileState) -> FileState {
-        // if not open for reading, remove the readable flag
-        if !self.mode.contains(FileMode::READ) {
-            state.remove(FileState::READABLE);
+    /// Align the pipe's state to the buffer state. For example if the buffer is both `READABLE` and
+    /// `WRITABLE`, and the pipe is only open in `READ` mode, the pipe's `READABLE` state will be
+    /// set and the `WRITABLE` state will be unchanged.
+    fn align_state_to_buffer(&mut self, buffer_state: BufferState, event_queue: &mut EventQueue) {
+        let mut mask = FileState::empty();
+        let mut file_state = FileState::empty();
+
+        // if the pipe is already closed, do nothing
+        if self.state.contains(FileState::CLOSED) {
+            return;
         }
 
-        // if not open for writing, remove the writable flag
-        if !self.mode.contains(FileMode::WRITE) {
-            state.remove(FileState::WRITABLE);
+        // only update the readable state if the file is open for reading
+        if self.mode.contains(FileMode::READ) {
+            mask.insert(FileState::READABLE);
+            // file is readable if the buffer is readable or there are no writers
+            if buffer_state.intersects(BufferState::READABLE | BufferState::NO_WRITERS) {
+                file_state.insert(FileState::READABLE);
```

Comment:
What happens when a read operation is performed on a pipe with no writers? Successful read of 0 bytes? Do we have a test case for this?

Reply:
Yeah it returns 0 (EOF). This should be covered by the two tests `test_read_after_write_close_with_empty_buffer` and `test_read_after_write_close_with_nonempty_buffer`.

Reply:
How did these tests pass before?

More generally if this change is fixing behavior (and not a pure refactor) then ideally there'd be a new test for the previously-broken behavior. (Or some existing test modified to catch why the previous behavior was wrong)

Reply:
The pipe behaviour was fine, but the unix socket behaviour was wrong. There aren't currently any tests for sending/receiving on connected unix sockets since shadow doesn't yet support `connect()` for unix sockets, and we can't add tests for this until we support `connect()`. This was just a preemptive bug fix, and should hopefully be tested by the unix socket tests when we add them.

Reply:
Ah, ok gotcha